File: OJ/leetcode/python/752_open_the_lock.py
from typing import List
from collections import deque

class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        candidates = [True for _ in range(10000)]
        for end in deadends:
            candidates[int(end)] = False
        if not candidates[0]:
            return -1
        target = int(target)
        if target == 0:
            return 0

        queue = deque()
        queue.appendleft((0, 0))
        candidates[0] = False
        depth = 0
        while len(queue) > 0:
            parent, depth = queue.pop()
            thousand = parent // 1000 * 1000
            hundred = parent // 100 * 100
            ten = parent // 10 * 10
            possible = [
                (parent - 1000) % 10000, 
                (parent + 1000) % 10000,
                (parent - thousand - 100) % 1000 + thousand, 
                (parent - thousand + 100) % 1000 + thousand,
                (parent - hundred - 10) % 100 + hundred, 
                (parent - hundred + 10) % 100 + hundred,
                (parent - ten - 1) % 10 + ten, 
                (parent - ten + 1) % 10 + ten
            ]
            for candidate in possible:
                if candidate == target:
                    return depth + 1
                if candidates[candidate]:
                    queue.appendleft((candidate, depth + 1))
                    candidates[candidate] = False
            depth += 1
        return -1



solution = Solution()

deadends = ["0201","0101","0102","1212","2002"]
target = "0202"
# ans = 6,  "0000" -> "1000" -> "1100" -> "1200" -> "1201" -> "1202" -> "0202"
print(solution.openLock(deadends, target))

deadends = ["8888"]
target = "0009"
print(solution.openLock(deadends, target))

deadends = ["8887","8889","8878","8898","8788","8988","7888","9888"]
target = "8888"
print(solution.openLock(deadends, target))

deadends = ["0000"]
target = "8888"
print(solution.openLock(deadends, target))

File: competitive_programming/leetcode/cpp/752_open_the_lock.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>
using namespace std;

class Solution {
private:
    inline int positiveModulo(int i, int m) {
        return (i % m + m) % m; 
    }
    vector<int> getCandidates(int cur) {
        int thousand = cur / 1000 * 1000;
        int hundred = cur / 100 * 100;
        int ten = cur / 10 * 10;
        return {
            positiveModulo(cur + 1000, 10000), 
            positiveModulo(cur - 1000, 10000),
            positiveModulo(cur - thousand + 100, 1000) + thousand,
            positiveModulo(cur - thousand - 100, 1000) + thousand,
            positiveModulo(cur - hundred + 10, 100) + hundred,
            positiveModulo(cur - hundred - 10, 100) + hundred,
            positiveModulo(cur - ten + 1, 10) + ten,
            positiveModulo(cur - ten - 1, 10) + ten
        };
    }
public:
    int openLock(vector<string>& deadends, string target) {
        vector<bool> valid(10000, true);
        for (string deadend : deadends) {
            valid[stoi(deadend)] = false;
        }
        if (!valid[0]) {
            return -1;
        }
        int ti = stoi(target);
        if (ti == 0) {
            return 0;
        }
        queue<pair<int, int>> q;
        q.push(make_pair(0, 0));
        valid[0] = false;
        pair<int, int> p;
        while (!q.empty()) {
            p = q.front();
            q.pop();
            int wheel = p.first;
            int depth = p.second;
            if (wheel == ti) {
                return depth;
            }
            vector<int> candidtes = getCandidates(wheel);
            for (int candidate : candidtes) {
                if (valid[candidate]) {
                    valid[candidate] = false;
                    q.push(make_pair(candidate, depth + 1));
                }
            }
        }
        return -1;
    }
};

int main() {
    Solution* solution = new Solution();
    vector<string> deadends;

    deadends = {"0201", "0101", "0102", "1212", "2002"};
    cout << solution->openLock(deadends, "0202") << "\n";

    deadends = {"8888"};
    cout << solution->openLock(deadends, "0009") << "\n";

    deadends = {"8887", "8889", "8878", "8898", "8788", "8988", "7888", "9888"};
    cout << solution->openLock(deadends, "8888") << "\n";

    deadends = {"0000"};
    cout << solution->openLock(deadends, "8888") << "\n";

    return 0;
}